core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,341 @@
+        inputStream.close();
+        return;
+      }
+      UfsInputStreamIds resources =
+          mFileIdToInputStreamIds.get(((CachedSeekableInputStream) inputStream).getFileId());
+      if (!resources.release(((CachedSeekableInputStream) inputStream).getResourceId())) {
+        LOG.debug("Close the expired input stream resource of {}",
+            ((CachedSeekableInputStream) inputStream).getResourceId());
+        // the input stream expired, close it
+        inputStream.close();
+      }
+    }
+  }
+
+  /**
+   * Invalidates an input stream from the cache.
+   *
+   * @param inputStream the cached input stream
+   * @throws IOException when the invalidated input stream fails to release
+   */
+  public void invalidate(CachedSeekableInputStream inputStream) throws IOException {
+    mUnderFileInputStreamCache.invalidate(inputStream.getResourceId());
+    release(inputStream);
+  }
+
+  /**
+   * Acquires an input stream. For seekable input streams, if there is an available input stream in
+   * the cache, reuse it and repositions the offset, otherwise the manager opens a new input stream.
+   *
+   * @param ufs the under file system
+   * @param path the path to the under storage file
+   * @param fileId the file id
+   * @param openOptions the open options
+   * @return the acquired input stream
+   * @throws IOException if the input stream fails to open
+   */
+  public InputStream acquire(UnderFileSystem ufs, String path, long fileId, OpenOptions openOptions)
+      throws IOException {
+    return acquire(ufs, path, fileId, openOptions, true);
+  }
+
+  /**
+   * Acquires an input stream. For seekable input streams, if there is an available input stream in
+   * the cache and reuse mode is specified, reuse it and repositions the offset, otherwise the
+   * manager opens a new input stream.
+   *
+   * @param ufs the under file system
+   * @param path the path to the under storage file
+   * @param fileId the file id
+   * @param openOptions the open options
+   * @param reuse true to reuse existing input stream, otherwise acquire a new stream
+   * @return the acquired input stream
+   * @throws IOException if the input stream fails to open
+   */
+  public InputStream acquire(UnderFileSystem ufs, String path, long fileId, OpenOptions openOptions,
+      boolean reuse) throws IOException {
+    if (!ufs.isSeekable() || !isCachingEnabled()) {
+      // not able to cache, always return a new input stream
+      return ufs.open(path, openOptions);
+    }
+
+    // explicit cache cleanup
+    mUnderFileInputStreamCache.cleanUp();
+
+    UfsInputStreamIds resources;
+    synchronized (mFileIdToInputStreamIds) {
+      if (mFileIdToInputStreamIds.containsKey(fileId)) {
+        resources = mFileIdToInputStreamIds.get(fileId);
+      } else {
+        resources = new UfsInputStreamIds();
+        mFileIdToInputStreamIds.put(fileId, resources);
+      }
+    }
+
+    synchronized (resources) {
```

Comment:
Would we be able to avoid this synchronization if `availableIds` was a `ConcurrentHashSet` (our impl)?

Reply:
I think it's not sufficient as the atomic operations on both `availableIds` and `inUseIds`